### ISSUE TITLE
Version 0.14.3 fails to compile against numpy-1.26.4

### ISSUE THREAD

**Issue by berndf** [statsmodels/statsmodels#9366]
#### Describe the bug

Cf. https://bugs.gentoo.org/930149

Here is the background:
https://numpy.org/devdocs/numpy_2_0_migration_guide.html#migration-c-descr

It appears that the migration advice given there was not heeded when making statsmodels compatible with numpy-2: You're supposed to do

#if NPY_ABI_VERSION < 0x02000000
  #define PyDataType_ELSIZE(descr) ((descr)->elsize)
#endif

This would have to be done for PyDataType_ALIGNMENT and others as well.

#### Code Sample, a copy-pastable example if possible

```python
# Your code here that produces the bug
# This example should be self-contained, and so not rely on external data.
# It should run in a fresh ipython session, and so include all relevant imports.
```
3.12.6

**Note**: As you can see, there are many issues on our GitHub tracker, so it is very possible that your issue has been posted before. Please check first before submitting so that we do not have to handle and close duplicates.

**Note**: Please be sure you are using the latest released version of `statsmodels`, or a recent build of `main`. If your problem has been fixed in an unreleased version, you might be able to use `main` until a new release occurs. 

**Note**: If you are using a released version, have you verified that the bug exists in the main branch of this repository? It helps the limited resources if we know problems exist in the current main branch so that they do not need to check whether the code sample produces a bug in the next release.

</details>

If the issue has not been resolved, please file it in the issue tracker.

#### Output of ``import statsmodels.api as sm; sm.show_versions()``

<details>

[paste the output of ``import statsmodels.api as sm; sm.show_versions()`` here below this line]

</details>

**Comment by berndf** [comment]
... or add npy2_compat.h ...

**Comment by bashtage** [comment]
This is expected. Statsmodels >= 0.14.2 requires NumPy 2.0.0 to compile. It will happily run against NumPy 1.22.3+.

**Comment by bashtage** [comment]
You also need a very recent Cython (I think 3.0.10+)

**Comment by bashtage** [comment]
Closing since this is as expected, and it appears to be a build configuration issue on Gentoo. We test against NumPy < 2, although building always requires NumPy 2.0.0+

**Comment by berndf** [comment]
> You also need a very recent Cython (I think 3.0.10+). The file with the error is coming from a cythonized pyx file, so I suspect that you have the wrong Cython.

Thanks for the clarifications!

Cython is 3.0.11 here.

As gentoo is a source-based distribution this means that 0.14.3 all by sudden, in a minor release, is numpy-2 only.
It actually cannot be too hard to allow compilation with a current numpy-1, along the lines of the compat document above, and several projects took that path. But that's your decision of course.

**Comment by bashtage** [comment]
I saw that you are using NumPy 1.26.  There is unfortunately no going back. The choice we faced was not having a NumPy 2 release until 0.15.0 which has no known release date or bumping 0.14.2 to support NumPy 2 (and require for compilation).

statsmodels is not NumPy 2 only.  It is NumPy 1.22.3+, although compilation requires NumPy 2. Could install NumPy 2 for compilation and then downgrade to a lower version. No idea is Gentoo supports this type of structure though. 

In full disclosure there isn't a lot to lose by sticking with with statsmodel 0.14.1. Almost all of the changes were what was needed to become NumPy 2 compatible. While there are a few fixes for recent pandas changes in .2 and .3, they are fairly modest. 0.14.1 also compiles with very old NumPy (IIRC possibly as old as 1.19). 

The structure we have is the same as all other NumPy 2 supporting projects. They require 2+ for compilation but have binary comparability with 1.22.3+.

Gentoo users can always compile from pip + sdist, or install the wheel if it is too much trouble.
